File: Graphs C++/13 - Articulation Point and Bridges/articulation point and bridges.cpp
#include<bits/stdc++.h>

using namespace std;

const int N = 1e5;

vector<int> gr[N];

int vis[N], disc[N], low[N], tme = 1;

vector<pair<int, int>> bridges;
set<int> arti_points;

void dfs(int cur, int par) {
	vis[cur] = 1;
	disc[cur] = low[cur] = tme++;
	int child = 0;
	for (auto x : gr[cur]) {

		if (!vis[x]) {
			dfs(x, cur);
			child++;
			// we know low and disc of x
			low[cur] = min(low[cur], low[x]);

			// bridges
			if (low[x] > disc[cur]) {
				bridges.push_back({cur, x});
			}

			// articulation points
			if (par != 0 && low[x] >= disc[cur]) {
				arti_points.insert(cur);
			}

		}
		else if (x != par) {
			// backedge
			low[cur] = min(low[cur], disc[x]);
		}
	}

	// root is an arti or not
	if (par == 0 && child > 1) {
		arti_points.insert(cur);
	}

	return;
}


int main()
{
	freopen("input.txt", "r", stdin);
	freopen("output.txt", "w", stdout);

	int n, m;
	cin >> n >> m;

	for (int i = 0; i < m; i++) {
		int x, y;
		cin >> x >> y;
		gr[x].push_back(y);
		gr[y].push_back(x);
	}

	dfs(1, 0);

	for (auto x : arti_points) cout << x << '\n';

	for (auto x : bridges) {
		cout << x.first << " " << x.second << '\n';
	}

	return 0;
}

File: Graphs Java/src/ArticulationPointAndBridges/ArticulationPointAndBridges.java
package ArticulationPointAndBridges;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

public class ArticulationPointAndBridges {

    static class Pair{
        int first, second;

        Pair(int a, int b){
            first = a;
            second = b;
        }
    }

    static final int N = (int) (1e5 + 1);

    static ArrayList<Integer>[] gr = new ArrayList[N];
    static Set<Integer> arti_points = new HashSet<>();
    static {
        for(int i = 0 ; i < N ; i++){
            gr[i] = new ArrayList<>();
        }
    }

    static int tme = 1;
    static boolean[] vis = new boolean[N];
    static int[] disc = new int[N];
    static int[] low = new int[N];

    static ArrayList<Pair> bridges = new ArrayList<>();

    static void dfs(int cur, int par) {
        vis[cur] = true;
        disc[cur] = low[cur] = tme++;
        int child = 0;
        for (int x : gr[cur]) {

            if (!vis[x]) {
                dfs(x, cur);
                child++;
                // we know low and disc of x
                low[cur] = Math.min(low[cur], low[x]);

                // bridges
                if (low[x] > disc[cur]) {
                    bridges.add(new Pair(cur, x));
                }

                // articulation points
                if (par != 0 && low[x] >= disc[cur]) {
                    arti_points.add(cur);
                }

            }
            else if (x != par) {
                // backedge
                low[cur] = Math.min(low[cur], disc[x]);
            }
        }

        // root is an arti or not
        if (par == 0 && child > 1) {
            arti_points.add(cur);
        }

    }


    public static void main(String[] args)
    {
        Scanner scn = new Scanner(System.in);
        int n = scn.nextInt(), m = scn.nextInt();

        for (int i = 0; i < m; i++) {
            int x = scn.nextInt(), y = scn.nextInt();
            gr[x].add(y);
            gr[y].add(x);

        }

        dfs(1, 0);

        for (int x : arti_points) System.out.println(x);

        for (Pair x : bridges) {
            System.out.println( x.first + " " + x.second );
        }

    }
}
